### DISCUSSION TITLE
Reading backwards through IEventReader

### DISCUSSION THREAD

**Discussion by joelving** [Eventuous/eventuous#103] [Q&A]
For some of my use cases I enumerate event streams backwards (e.g. get latest X events), which is perfectly supported by Eventstore through it's Direction enum. The IEventReader.ReadEvents method unfortunately does not allow this to be passed in, meaning that reads through the interface can only happen forward. Without a method to retrieve the latest position, I believe this leaves the user with no way to enumerate the stream backwards without reading the entire stream.

Am I missing a way to accomplish this without tying myself to Eventstoredb?

I considered fetching the position of the latest event in the stream, but the IEventReader does not support this either.

**Comment by alexeyzimarev** [comment]
The issue there was that not every store is able to read backwards, and the function itself wasn't used in Eventuous. It forced me to have some implementations to throw a "not implemented" exception.

The `EsdbEventStore`, for example, has the `ReadStreamBackwards` still. I am not sure if this should be a part of the abstraction. Or maybe it should be a completely separate interface.

**Reply by joelving** [reply]
> The issue there was that not every store is able to read backwards

This is surprising to me. Would it not just require being able to sort on an insertion marker (sequential Id, timestamp, rowversion, whatever)?

Anyways, what I did end up doing was as you suggest: Add a different interface, which derives from IEventReader. That does lead to a bit of code duplication when augmenting e.g. `EsdbEventStore`, but so be it. Definitely better than having implementations throw (though maybe that could be mitigated by CanReadBackwards flag?).

**Reply by alexeyzimarev** [reply]
I probably will consider getting it back when I start working on the embedded snapshotting functionality, as this will be required.
